Anchor optim parallel

mlampros opened the pull request:
First of all thanks for making this repository publicly available.

This PR,

* adds parallelization (multiple threads in the 'scipy.optimize.differential_evolution' function). For instance - depending on the task - using 8 cores might be 4 to 5 times faster than using a single thread.
* restructures the 'anchor_optimization' Python package so that it can be installed using pip (added the __ init __.py files, 'anchor_optimization' and 'utils' directories)
* adds the entry point 'anchor-optim' in the 'setup.py' file in the same way the "keras-retinanet" works (from the command line)
* adds the 'mutation', 'threads' and 'verbose' parameters
* splits the code in two files ('optimize_anchors.py' and 'optimize_anchors_argparse.py')
* updates the README.md file

Important notes,

* a scipy version of 1.2.0 is required for [multiprocessing](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html)
* when the number of threads is > 1 the 'updating' parameter is set to 'deferred' by default (documentation of "scipy.optimize.differential_evolution()")
* I observed that the Python 'anchors_optim()' function might give incorrect results from within a Python console (it's recommended to be used from the command line only using the entry point 'anchor-optim')

mlampros committed:
parallelized version and restructure code

mlampros committed:
parallelized version and restructure code

martinzlocha commented:
Hi, thank you very much for your contribution, I will have a more detailed look at the code during the weekend. In the meantime could you get it a bit closer to PEP-8 compliance? I would like to keep the code as clean as possible.

mlampros committed:
update the code based on PEP-8

mlampros commented:
Hi, I updated the code as you suggested. I used the [autopep8](https://github.com/hhatto/autopep8) package for this purpose,

```bash
autopep8 --in-place --aggressive --aggressive --recursive /anchor-optimization
```

with the only difference that I moved the 

```bash
warnings.filterwarnings('ignore', category=FutureWarning)
```

line at the top of the file. 
Let me know if there's something else that has to be modified.

mlampros committed:
update setup.py for tensorflow 'cpu' or 'gpu' version

mlampros commented:
Moreover, I've added a code snippet in the 'setup.py' file to check if any of the tensorflow 'cpu' or 'gpu' versions exist (to avoid re-installation).

martinzlocha commented on README.md:
```diff
@@ -8,15 +8,15 @@ For questions and discussion join the [Keras Slack](https://keras-slack-autojoin
 1. Clone this repository
-1. `pip install .`
-1. `python setup.py build_ext --inplace`
+2. `pip install .`
```
Just an FYI, this change isn't really needed, markdown will automatically change the numbers so that they are increasing. So keeping it all as 1s make it easier to add new elements to the list.

mlampros replied:
Thanks for making me aware, I didn't know that. I'll fix this.

martinzlocha commented on README.md:
```diff
@@ -8,15 +8,15 @@ For questions and discussion join the [Keras Slack](https://keras-slack-autojoin
 
 1. Define your own dataset in a csv format, for more information follow the guide in [keras-retinanet](https://github.com/fizyr/keras-retinanet#csv-datasets).
-1. Run `python optimize_anchors.py PATH_TO_CSV` 
+2. Run `anchor-optim PATH_TO_CSV` 
```
Could we use a `anchor-optimization` instead, its just saving a few characters.

mlampros replied:
Yes, I'll modify the command from 'anchor-optim' to 'anchor-optimization'

martinzlocha commented on anchor_optimization/optimize_anchors.py:
```diff
@@ -0,0 +1,347 @@
+
+import warnings
+warnings.filterwarnings('ignore', category=FutureWarning)
+# warnings.simplefilter("ignore")
```
Please remove commented out code.

mlampros replied:
I'll do this.

martinzlocha commented on anchor_optimization/optimize_anchors.py:
```diff
@@ -0,0 +1,347 @@
+                                                   popsize=popsize,
+                                                   seed=seed)
+
+    # ------------------------------------------------------------------------------------------------------ using the 'lambda' function gives an error because 'lambdas' are not pickleable and the 'workers' parameter requires that 'func' is pickleable [SEE: https://stackoverflow.com/questions/25348532/can-python-pickle-lambda-functions ]
```
I think this comment could be removed, it takes up a lot of space but doesn't really add much value.

mlampros replied:
I'll do this.

martinzlocha commented on anchor_optimization/optimize_anchors.py:
```diff
@@ -0,0 +1,347 @@
+
+    result = scipy.optimize.differential_evolution(func=average_overlap,
+                                                   args=ARGS,
+                                                   # pass the *args as a tuple
```
Put the comment into a single line and move it above the args parameter.

mlampros replied:
thanks, I missed this one.

martinzlocha commented on anchor_optimization/optimize_anchors.py:
```diff
@@ -0,0 +1,347 @@
+
+
+# ***************************************************************************************************************************************************
+# !!!! the python "anchors_optim" function is meant to be used from the command line (from within a Python console it gives incorrect results)  !!!!
```
This is a weird statement, why does it not give correct results? Also can you use docstring comments? https://www.python.org/dev/peps/pep-0257/#id17

mlampros replied:
This is something I can't figure out. Initially I thought that it might had to do with the fact that I had 'keras-retinanet' both in Python2 and Python3 in my OS installed. However, this is not the case because I run the function today in a docker environment where only Python3 is installed and I get different results although using the same seed,

```R

# from the command-line 

State: 0.10112
Ratios: [0.471, 0.731, 1.0, 1.367, 2.123]
Scales: [0.541, 0.655, 0.781, 0.913]
Number of labels that dont have any matching anchor: 2

# from within Python

State: 0.21237
Ratios: [0.449, 0.728, 1.0, 1.373, 2.226]
Scales: [0.4, 0.469, 0.549, 0.656]
Number of labels that dont have any matching anchor: 1653

```
Obviously, from within Python gives the wrong output. I run the same script a couple of days ago also using a single thread and I received similar results.
The code is actually the same as before this PR, just wrapped in a function and exposed to Python (previously this wasn't possible as it was run only from the command line). What do you think?

martinzlocha replied:
Hmm not sure, I will try to have a look at this but I am a bit short on time so will accept this PR as is.

martinzlocha commented on anchor_optimization/optimize_anchors.py:
```diff
@@ -0,0 +1,347 @@
+# !!!! the python "anchors_optim" function is meant to be used from the command line (from within a Python console it gives incorrect results)  !!!!
+# ***************************************************************************************************************************************************
+
+def anchors_optim(annotations,
```
Lets change this to anchors_optimize.

mlampros replied:
I'll modify this too.

martinzlocha commented on anchor_optimization/optimize_anchors_argparse.py:
```diff
@@ -0,0 +1,127 @@
+
+import warnings
+warnings.filterwarnings('ignore', category=FutureWarning)
+# warnings.simplefilter("ignore")
```
This can be removed.

martinzlocha reviewed (changes requested):
Sorry for making you wait, just left a few minor stylistic comments. Let me know if anything is unclear.

mlampros committed:
updated 'anchor-optimization' files

martinzlocha commented:
Thank you for the PR, hopefully others find it valuable.
